Source/WebCore/platform/graphics/skia/GraphicsContextSkia.cpp

```diff
@@ -901,9 +901,21 @@ void GraphicsContextSkia::drawPattern(NativeImage& nativeImage, const FloatRect&
     SkPaint paint = createFillPaint();
     paint.setBlendMode(toSkiaBlendMode(options.compositeOperator(), options.blendMode()));
 
-    if (spacing.isZero() && tileRect.size() == nativeImage.size())
-        paint.setShader(image->makeShader(SkTileMode::kRepeat, SkTileMode::kRepeat, samplingOptions, &shaderMatrix));
-    else {
+    if (spacing.isZero() && tileRect.size() == nativeImage.size()) {
+        // Check whether we're sampling the pattern beyond the image size. If this is the case, we need to set the repeat
+        // flag when sampling. Otherwise we use the clamp flag. This is done to avoid a situation where the pattern is scaled
+        // to fit perfectly the destinationRect, but if we use the repeat flag in that case the edges are wrong because the
+        // scaling interpolation is using pixels from the other end of the image.
+        SkMatrix inverse;
+        if (shaderMatrix.invert(&inverse)) {
+            SkRect imageSampledRect;
+            inverse.mapRect(&imageSampledRect, SkRect::MakeXYWH(destRect.x(), destRect.y(), destRect.width(), destRect.height()));
+            bool repeatX = imageSampledRect.x() < 0 || std::trunc(imageSampledRect.width()) > nativeImage.size().width();
+            bool repeatY = imageSampledRect.y() < 0 || std::trunc(imageSampledRect.height()) > nativeImage.size().height();
+            paint.setShader(image->makeShader(repeatX ? SkTileMode::kRepeat : SkTileMode::kClamp, repeatY ? SkTileMode::kRepeat : SkTileMode::kClamp, samplingOptions, &shaderMatrix));
+        } else
+            paint.setShader(image->makeShader(SkTileMode::kRepeat, SkTileMode::kRepeat, samplingOptions, &shaderMatrix));
```

Comment:
Maybe we can initialize repeat X/Y as true, set it to false if needed inside the if and keep a single call to setShader

Reply:
Sure! I didn't worry too much about that yet, as I wanted to check the test results first, but I'll handle that in the final version.